File: backend/app/schemas/shift_request.py
"""shift_requests に関する Pydantic スキーマ。"""
from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field, model_validator


class ShiftRequestCreate(BaseModel):
    """希望の提出リクエスト。"""

    work_date: date
    request_type: Literal["day_off", "work_preference"]
    shift_type_id: int | None = None
    note: str | None = Field(default=None, max_length=255)

    @model_validator(mode="after")
    def check_shift_type_for_preference(self) -> "ShiftRequestCreate":
        """勤務希望のときは希望する勤務区分を必須にする。"""
        if self.request_type == "work_preference" and self.shift_type_id is None:
            raise ValueError("勤務希望の場合は勤務区分を選択してください")
        return self


class ShiftRequestRead(BaseModel):
    """API レスポンス用の希望1件。管理者の一覧表示用に職員名も同梱する。"""

    id: int
    user_id: int
    user_name: str
    work_date: date
    shift_type_id: int | None
    request_type: str
    status: str
    note: str | None
    created_at: datetime


class ShiftRequestStatusUpdate(BaseModel):
    """希望の承認/却下リクエスト(管理者のみ)。"""

    status: Literal["accepted", "rejected"]

File: frontend/src/types/shiftRequest.ts
/** 希望の種類。day_off = 休み希望、work_preference = 勤務希望 */
export type RequestType = 'day_off' | 'work_preference'

/** 希望の状態。pending = 審査中 */
export type RequestStatus = 'pending' | 'accepted' | 'rejected'

/** バックエンドの ShiftRequestRead スキーマに対応する型。 */
export interface ShiftRequest {
  id: number
  user_id: number
  user_name: string
  /** "YYYY-MM-DD" 形式 */
  work_date: string
  shift_type_id: number | null
  request_type: RequestType
  status: RequestStatus
  note: string | null
  created_at: string
}

/** 希望の提出リクエストボディ。 */
export interface ShiftRequestInput {
  work_date: string
  request_type: RequestType
  shift_type_id: number | null
  note?: string
}
